velox/connectors/hive/iceberg/DeleteFileReader.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/connectors/hive/FileHandle.h"
+#include "velox/connectors/hive/HiveDataSource.h"
+
+#include <folly/Executor.h>
+#include <memory>
+
+namespace facebook::velox::cache {
+class AsyncDataCache;
+}
+
+namespace facebook::velox::connector {
+class ConnectorQueryCtx;
+}
+
+namespace facebook::velox::core {
+class ExpressionEvaluator;
+}
+
+namespace facebook::velox::connector::hive::iceberg {
+
+class IcebergDeleteFile;
+
+class DeleteFileReader {
+ public:
+  DeleteFileReader(
+      const IcebergDeleteFile& deleteFile,
+      const std::string& baseFilePath,
+      FileHandleFactory* fileHandleFactory,
+      folly::Executor* executor,
+      ConnectorQueryCtx* connectorQueryCtx,
+      uint64_t splitOffset,
+      const std::string& connectorId);
+
+  void readDeletePositions(
+      uint64_t baseReadOffset,
+      uint64_t size,
+      int8_t* deleteBitmap);
+
+  bool endOfFile();
+
+ private:
+  void createPositionalDeleteDataSource(
+      const IcebergDeleteFile& deleteFile,
+      const std::string& connectorId);
+
+  void readDeletePositionsVec(
+      uint64_t baseReadOffset,
+      int64_t rowNumberUpperBound,
+      int8_t* deleteBitmap);
+
+  bool readFinishedForBatch(int64_t rowNumberUpperBound);
+
+  const IcebergDeleteFile& deleteFile_;
+  const std::string& baseFilePath_;
+  FileHandleFactory* const fileHandleFactory_;
+  folly::Executor* const executor_;
+  ConnectorQueryCtx* const connectorQueryCtx_;
+  uint64_t splitOffset_;
+
+  std::unique_ptr<HiveDataSource> deleteDataSource_;
```

Comment:
@yingsu00 Please use a `dwio::common::Reader` instead of the whole data source here.   There are a lot of stuff you don't need in a delta file reader: column manipulation, remaining filter, split prefetch, etc.

Reply:
@Yuhta Thanks for reviewing. Yes I can do that. I'll move the util functions like makeScanSpec, getColumnName, etc to a new HiveReaderUtil class, so that they can be shared by the HiveDataSource and the DeleteFileReader.

Reply:
That sounds good

Reply:
@Yuhta For positional deletes, we will need to build subfield filters on the path column, as an Iceberg positional delete file can contain deleted rows in multiple base files. Will also add filter on the delete row column, since it's sorted and the lower/upperbound are known. 
For equality deletes, not every column needs to be read sometimes.

Reply:
You don't need `makeScanSpec` for these, just construct the `ScanSpec` yourselves (see below code example), it is very small and controllable.  I did that for metalake.  `makeScanSpec` is more for translating the `SubfieldFilters` and prunings from Presto representation.

```c++
    auto spec = std::make_shared<common::ScanSpec>("<root>");
    spec->addAllChildFields(*rowType);
    auto* pathSpec = spec->childByName("path");
    pathSpec->setFilter(std::make_unique<common::BytesValues>(
        std::vector<std::string>{path}, false));
```